etcdserver/security/security.go

```diff
@@ -0,0 +1,381 @@
+
+func mergeErr(s string, v ...interface{}) MergeError {
+	return MergeError{fmt.Sprintf(s, v...)}
+}
+
+func NewStore(server doer, timeout time.Duration) *Store {
+	s := &Store{
+		server:  server,
+		timeout: timeout,
+	}
+	s.enabled = s.detectSecurity()
+	return s
+}
+
+func (s *Store) AllUsers() ([]string, error) {
+	resp, err := s.requestResource("/users/", false)
+	if err != nil {
+		return nil, err
+	}
+	var nodes []string
+	for _, n := range resp.Event.Node.Nodes {
+		_, user := path.Split(n.Key)
+		nodes = append(nodes, user)
+	}
+	sort.Strings(nodes)
+	return nodes, nil
+}
+
+func (s *Store) GetUser(name string) (User, error) {
+	resp, err := s.requestResource("/users/"+name, false)
+	if err != nil {
+		return User{}, err
+	}
+	var u User
+	err = json.Unmarshal([]byte(*resp.Event.Node.Value), &u)
+	if err != nil {
+		return u, err
+	}
+
+	return u, nil
+}
+
+func (s *Store) CreateUser(user User) error {
+	if user.User == "root" {
+		return mergeErr("Cannot create root user; enable security to set root.")
+	}
+	return s.createUserInternal(user)
+}
+
+func (s *Store) createUserInternal(user User) error {
+	hash, err := bcrypt.GenerateFromPassword([]byte(user.Password), bcrypt.DefaultCost)
+	if err != nil {
+		return err
+	}
+	user.Password = string(hash)
+
+	_, err = s.createResource("/users/"+user.User, user)
+	return err
+}
+
+func (s *Store) DeleteUser(name string) error {
+	if name == "root" {
+		return mergeErr("Can't delete root user; disable security instead.")
+	}
+	_, err := s.deleteResource("/users/" + name)
+	return err
+}
+
+func (s *Store) UpdateUser(user User) (User, error) {
+	old, err := s.GetUser(user.User)
+	if err != nil {
+		return old, err
+	}
+	newUser, err := old.Merge(user)
+	if err != nil {
+		return old, err
+	}
+	_, err = s.updateResource("/users/"+user.User, newUser)
+	if err != nil {
+		return newUser, err
+	}
+	return newUser, nil
+}
+
+func (s *Store) AllRoles() ([]string, error) {
+	resp, err := s.requestResource("/roles/", false)
+	if err != nil {
+		return nil, err
+	}
+	var nodes []string
+	for _, n := range resp.Event.Node.Nodes {
+		_, role := path.Split(n.Key)
+		nodes = append(nodes, role)
+	}
+	sort.Strings(nodes)
+	return nodes, nil
+}
+
+func (s *Store) GetRole(name string) (Role, error) {
+	// TODO(barakmich): Possibly add a cache
+	resp, err := s.requestResource("/roles/"+name, false)
+	if err != nil {
+		return Role{}, err
+	}
+	var r Role
+	err = json.Unmarshal([]byte(*resp.Event.Node.Value), &r)
+	if err != nil {
+		return r, err
+	}
+
+	return r, nil
+}
+
+func (s *Store) CreateRole(role Role) error {
+	_, err := s.createResource("/roles/"+role.Role, role)
+	return err
+}
+
+func (s *Store) DeleteRole(name string) error {
+	_, err := s.deleteResource("/roles/" + name)
+	return err
+}
+
+func (s *Store) UpdateRole(role Role) (Role, error) {
+	old, err := s.GetRole(role.Role)
+	if err != nil {
+		return old, err
+	}
+	newRole, err := old.Merge(role)
```

Comment:
shall we check the existence of the role before granting it? now i can grant non-existent role to a user.